### PR TITLE
Support Phi-4 Multi-Modal (text + vision only)

### PR DESCRIPTION
**Pull request by lifuhuang**
## Motivation

Support Phi4-MM model with text + vision.

## Modifications

This change introduced the basic text + image support. 
image

It's worth noting that the current MMMU run (without LoRA) is lower than advertised because Phi4MM relies on LoRA for full image understanding capabilities. However, LoRA support requires refactoring / generalizing the existing SGL LoRA handling, which will hopefully be addressed in this separate PR: https://github.com/sgl-project/sglang/pull/6585

Example: degraded image understanding without LoRA (MMMU is only 38). As comparison in our local branch (https://github.com/sgl-project/sglang/pull/6585) with LoRA, MMMU is boosted to ~50:
image

TODO in this PR:

- [x] add unit tests
- [x] clean-up styling issues

TODO in follow-up PR (ordered by priority):
1. Precomputed feature support.
2. LoRA support (required for multi-image understanding)
3. SGLang LoRA compatibility with CUDA Graph and Radix Attention
4. Refactor SGL MM processor logic support for support the original token variable image token (e.g., <image_1>)
5. perf optimization
6. audio support
7. pipeline parallelism support

Tracked in https://github.com/sgl-project/sglang/issues/6544

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by zhaochenyang20** [comment]
@mickqian @yizhang2077

File: python/sglang/srt/models/phi4mmvllm.py [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        forward_batch: ForwardBatch,
+        **kwargs: object,
+    ) -> torch.Tensor:
+        embed_tokens = self.model.embed_tokens
+        try:
+            if (
```
Comment (mickqian) [inline python/sglang/srt/models/phi4mmvllm.py]: Could you adapt to `general_mm_embed_routine`?

**Comment by mickqian** [comment]
Better to be merged after #4969 , due to some change to the omni model processing and testing

File: python/sglang/srt/models/phi4mmvllm.py (hunk 1 above)
Comment (lifuhuang) [reply python/sglang/srt/models/phi4mmvllm.py]: Yes, just updated in the latest iteration.

**Comment by lifuhuang** [comment]
> Better to be merged after #4969 , due to some change to the omni model processing and testing

Hi @mickqian, thank you so much for reviewing my PR :)

Can you share more details about the concern you have such that I can test them locally? JFYI, I was able to merge your branch mickqian:qwen2.5-omni locally without conflict and get a green TestOpenAIVisionServer run for phi4mm.

**Comment by mickqian** [comment]
> Can you share more details

~It's mostly that, for omni models, there's a new `TestOpenaiOmniServer`. And yes, you can cherry-pick it.~

I just noticed audio input is not supported this time.

File: python/sglang/srt/conversation.py [hunk 2]
```diff
@@ -661,6 +661,21 @@ def generate_chat_conv(
...
+    Conversation(
+        name="phi-4-mm",
+        system_message="You are a helpful language and vision assistant. You are able to understand the visual content that the user provides, and assist the user with a variety of tasks using natural language.",
+        system_template="<|system|>{system_message}<|end|>",
+        roles=("<|user|>", "<|assistant|>"),
+        sep_style=SeparatorStyle.NO_COLON_SINGLE,
+        sep="<|end|>",
+        stop_str="<|end|>",
+        image_token="<|endoftext10|>",
+        # image_token="<|image_{index}|>",
```
Comment (mickqian) [inline python/sglang/srt/conversation.py]: redundant

File: test/srt/test_vision_openai_server_b.py [hunk 3]
```diff
@@ -196,5 +196,31 @@ def test_video_chat_completion(self):
...
+                "--mem-fraction-static",
+                "0.75",
+            ],
+        )
+        cls.base_url += "/v1"
+
+    def test_video_chat_completion(self):
+        pass
+
+    def test_multi_images_chat_completion(self):
```
Comment (mickqian) [inline test/srt/test_vision_openai_server_b.py:220]: May i ask what's the problem here?

File: python/sglang/srt/models/phi4mmvllm.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+        image_attention_mask: num_images x num_crops x 32 x 32
+        output: (num_images, num_img_tokens, hidden_size)
+        """
+
+        # eg
+        # pixel_values: torch.Size([1, 7, 3, 448, 448])
+        # image_sizes: tensor([[ 896, 1344]], device='cuda:0')
+        # output: torch.Size([1, 1841, 3072])
+
+        if isinstance(self.img_projection, nn.Sequential):
```
Comment (mickqian) [inline python/sglang/srt/models/phi4mmvllm.py]: nit: `next(m.parameters()).device`

File: test/srt/test_vision_openai_server_b.py (hunk 3 above)
Comment (lifuhuang) [reply test/srt/test_vision_openai_server_b.py:220]: It generates much lower quality results than single image (e.g., understands one image correctly but then hallucinates with the other one), I tested in vLLM without LoRA and observed similar behavior. I am working on a quick follow-up PR for LoRA support.

**Review by mickqian — APPROVED** [review]

**Review by zhyncs — APPROVED** [review]
